web/app_dev.php

```diff
@@ -11,16 +11,36 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\Debug\Debug;
 
-// This check prevents access to debug front controllers that are deployed by accident to production servers.
-// Feel free to remove this, extend it, or make something more sophisticated.
+// Access to debug front controllers must only allowed from localhost or with username and password.
+// FIXME: add link to documentation that explains what to do.
+$accessKey = '';
+
 if (isset($_SERVER['HTTP_CLIENT_IP'])
     || isset($_SERVER['HTTP_X_FORWARDED_FOR'])
     || !(in_array(@$_SERVER['REMOTE_ADDR'], array('127.0.0.1', 'fe80::1', '::1')) || php_sapi_name() === 'cli-server')
 ) {
-    header('HTTP/1.0 403 Forbidden');
-    exit('You are not allowed to access this file. Check ' . basename(__FILE__) . ' for more information.');
+    list($username, $password) = explode(':', $accessKey);
+
+    if ('' === $username || '' === $password) {
+        header('HTTP/1.0 403 Forbidden');
+        exit('You are not allowed to access this file. Check ' . basename(__FILE__) . ' for more information.');
+
+    } elseif (!isset($_SERVER['PHP_AUTH_USER'])
+        || !isset($_SERVER['PHP_AUTH_PW'])
+        || hash('sha512', $_SERVER['PHP_AUTH_USER']) !== $username
+        || hash('sha512', $_SERVER['PHP_AUTH_PW']) !== $password
+    ) {
+        header('WWW-Authenticate: Basic realm="Contao Debug"');
+        header('HTTP/1.0 401 Unauthorized');
+        exit('You are not allowed to access this file. Check ' . basename(__FILE__) . ' for more information.');
+    }
+
+    unset($username, $password, $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW']);
```

Comment:
I do not like the approach of basic auth in general as digest would be more secure but I am with @leofeyer on the case that we MUST NOT tamper with the contents of `$_SERVER`.

I also am unsure if .htaccess hacks are still needed in cgi environments for PHP.

Further reading: http://php.net/manual/en/features.http-auth.php (example implementation of digest auth included). Also read the comments for some cgi environment hacks.